sql/parser/builtins.go

```diff
@@ -856,3 +870,48 @@ func round(x float64, n int64) (Datum, error) {
 	y, err := strconv.ParseFloat(strconv.FormatFloat(x, 'f', int(n), b), b)
 	return DFloat(y), err
 }
+
+var uniqueIDState struct {
+	sync.Mutex
+	millis uint64
+	rand   uint32
+}
+
+func generateUniqueID(nodeID uint32) DBytes {
+	// Unique IDs are composed of the current time in milliseconds, a 31-bit
+	// random number and a 32-bit node-id. If two unique IDs are generated within
+	// the same millisecond on a server, the random number is incremented instead
+	// of being generated fresh. This ensures that unique ID generation is
+	// monotonic on a single server and k-sorted across servers.
+	//
+	// Note that rand.Int31 returns a 32-bit positive signed integer. We cast
+	// this to a uint32. There are at least 1<<31 increments we can perform on
+	// this number before it wraps to 0, so such wrapping is not a concern within
+	// a given millisecond.
+	//
+	// TODO(pmattis): We could squeeze a bit more space out of the encoding. For
+	// example, we could limit the node-id to 16-bits and the random component to
+	// 16-bits and handle overflow of the rand value by incrementing the
+	// millisecond component.
+	//
+	// TODO(pmattis): Do we have to worry about persisting the milliseconds value
+	// periodically to avoid the clock ever going backwards (e.g. due to NTP
+	// adjustment)?
+	millis := uint64(time.Now().UnixNano() / int64(time.Millisecond))
+	uniqueIDState.Lock()
+	if millis <= uniqueIDState.millis {
+		millis = uniqueIDState.millis
+		uniqueIDState.rand++
+	} else {
+		uniqueIDState.millis = millis
+		uniqueIDState.rand = uint32(rand.Int31())
```

Comment:
I'm not entirely sure about the randomness either. I was initially following in the footsteps of Twitter's snowflake ID, though clearly we've already diverged from there. The possibilities here are: 
* `timestamp+random+nodeID` - guaranteed unique.
* `timestamp+counter+nodeID` - guaranteed unique, monotonic on a single node.
* `timestamp+random` - probabilistically unique. also makes the encoded ID slightly larger.
* `timestamp+nodeID+counter` - guaranteed unique, monotonic and sequential on a single node within the timestamp granularity.

The advantage to per-node monotonicity is that the insertion of multiple rows into a table will behave as expected:

```
  INSERT INTO t VALUES (unique_id(), "hello"), (unique_id(), "world")
```

Without the per-node monotonicity guarantee, the ordering of these rows would be random. We could definitely document this, but it seemed too surprising when I thought about it.